libs/utils/src/leaky_bucket.rs

```diff
@@ -0,0 +1,225 @@
+use std::time::Duration;
+
+use tokio::time::Instant;
+
+pub struct LeakyBucketConfig {
+    /// Leaky buckets can drain at a fixed interval rate.
+    /// We track all times as durations since this epoch so we can round down.
+    pub epoch: Instant,
+
+    /// How frequently we drain the bucket.
+    /// If equal to 0, we drain continuously over time.
+    /// If greater than 0, we drain at fixed intervals.
+    pub drain_interval: Duration,
+
+    /// "time cost" of a single request unit.
+    /// should loosely represent how long it takes to handle a request unit in active resource time.
+    pub cost: Duration,
+
+    /// total size of the bucket
+    pub bucket_width: Duration,
+}
+
+impl LeakyBucketConfig {
+    fn prev_multiple_of_drain(&self, mut dur: Duration) -> Duration {
+        if self.drain_interval > Duration::ZERO {
+            let n = dur.div_duration_f64(self.drain_interval).floor();
+            dur = self.drain_interval.mul_f64(n);
+        }
+        dur
+    }
+
+    fn next_multiple_of_drain(&self, mut dur: Duration) -> Duration {
+        if self.drain_interval > Duration::ZERO {
+            let n = dur.div_duration_f64(self.drain_interval).ceil();
+            dur = self.drain_interval.mul_f64(n);
+        }
+        dur
+    }
+}
+
+pub struct LeakyBucketState {
+    /// Bucket is represented by `start..end` where `end = epoch + end` and `start = end - config.bucket_width`.
+    ///
+    /// At any given time, `end - now` represents the number of tokens in the bucket, multiplied by the "time_cost".
+    /// Adding `n` tokens to the bucket is done by moving `end` forward by `n * config.time_cost`.
+    /// If `now < start`, the bucket is considered filled and cannot accept any more tokens.
+    /// Draining the bucket will happen naturally as `now` moves forward.
+    ///
+    /// Let `n` be some "time cost" for the request,
+    /// If now is after end, the bucket is empty and the end is reset to now,
+    /// If now is within the `bucket window + n`, we are within time budget.
+    /// If now is before the `bucket window + n`, we have run out of budget.
+    ///
+    /// This is inspired by the generic cell rate algorithm (GCRA) and works
+    /// exactly the same as a leaky-bucket.
+    pub end: Duration,
```

Comment:
I feel like this doc comment should be at the top of the module. Without it, the module is a lot harder to understand. Put it in front of the reader's eyes before they dive into the code.

Also, it's not explained anywhere where `start` comes from.

Mention of GCRA and a link to the wikipedia page should IMO be in the first pargagraph, not the last.